sonar-application/src/main/java/org/sonar/application/Cluster.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.application;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hazelcast.cluster.ClusterState;
+import com.hazelcast.config.Config;
+import com.hazelcast.config.JoinConfig;
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import java.util.Arrays;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Manager for the cluster communication between Main Processes
+ */
+public class Cluster implements AutoCloseable {
```

Comment:
I provided a code sample where I suggested to add a try-with-resource to the main method (and assumed that remove the need to handle the stop explicitly in the code)

you didn't go in that direction and put the close directly in the `stopAsync`, so the `AutoCloseable` is not required anymore. I didn't consider this solution at the time.

But now, considering the subject again, I think you should keep the AutoCloseable and use try-with-resource around the call to `app.start()` as I suggested. It's an elegant solution, but more importantly, it is a more reliable one.

`app.start()` is blocking because it calls `monitor.awaitTermination()` which is blocking. `stopAsync` is stopping SQ because it calls `monitor.stop()` which will unblock `monitor.awaitTermination()`.
But assume that `monitor.awaitTermination()` unblocks for another reason (exception or error), then `stopAsync` will never be called and hazelcast won't be stop cleanly.

Last killing reason to not call stop in `stopAsync`: it's not used except in one location in tests!
(it's a left over of the time where the Stopping thread was part of App before it became part of Monitor)